ecosystem/gstreamer_plugin/gst_mtl_st20p_rx.c

```diff
@@ -0,0 +1,658 @@
+      g_value_set_string(value, src->pixel_format);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+      break;
+  }
+}
+
+/*
+ * Create MTL session rx handle and initialize the session with the parameters
+ * from caps negotiated by the pipeline.
+ */
+
+static gboolean gst_mtl_st20p_rx_negotiate(GstBaseSrc* basesrc) {
+  GstVideoInfo* info;
+  Gst_Mtl_St20p_Rx* src = GST_MTL_ST20P_RX(basesrc);
+  struct st20p_rx_ops* ops_rx = &src->ops_rx;
+  gint ret;
+  GstCaps* caps;
+
+  info = gst_video_info_new();
+
+  /*
+   * Convert boolean interlaced value to integer,
+   * as only basic interlaced mode (1) is supported.
+   */
+  info->interlace_mode = src->interlaced;
+
+  switch (ops_rx->output_fmt) {
+    case ST_FRAME_FMT_V210:
+      info->finfo = gst_video_format_get_info(GST_VIDEO_FORMAT_v210);
+      break;
+    case ST20_FMT_YUV_420_10BIT:
+      info->finfo = gst_video_format_get_info(GST_VIDEO_FORMAT_I422_10LE);
+      break;
+    default:
+      GST_ERROR("Unsupported pixel format");
+      return FALSE;
+  }
+
+  caps = gst_caps_new_simple("video/x-raw", "format", G_TYPE_STRING, "v210", "width",
+                             G_TYPE_INT, info->width, "height", G_TYPE_INT, info->height,
+                             "framerate", GST_TYPE_FRACTION, info->fps_n, 1,
+                             "interlace-mode", G_TYPE_BOOLEAN, src->interlaced, NULL);
+
+  if (!caps) caps = gst_pad_get_pad_template_caps(GST_BASE_SRC_PAD(basesrc));
+
+  if (gst_caps_is_empty(caps)) {
+    gst_caps_unref(caps);
+    return FALSE;
+  }
+
+  ret = gst_pad_set_caps(GST_BASE_SRC_PAD(basesrc), caps);
+  gst_caps_unref(caps);
+  if (!ret) {
+    GST_ERROR("Failed to set caps");
+    return FALSE;
+  }
+
+  gst_video_info_free(info);
+  return TRUE;
+}
+
+static GstFlowReturn gst_mtl_st20p_rx_create(GstBaseSrc* basesrc, guint64 offset,
+                                             guint length, GstBuffer** buffer) {
+  GstBuffer* buf;
+  Gst_Mtl_St20p_Rx* src = GST_MTL_ST20P_RX(basesrc);
+  struct st_frame* frame;
+  GstMapInfo dest_info;
+  gint ret;
+  gsize fill_size;
+
+  buf = gst_buffer_new_allocate(NULL, src->frame_size + 1, NULL);
+  if (!buf) {
+    GST_ERROR("Failed to allocate buffer");
+    return GST_FLOW_ERROR;
+  }
+
+  *buffer = buf;
+
+  GST_OBJECT_LOCK(src);
+
+  for (int i = 0; i < src->retry_frame; i++) {
+    frame = st20p_rx_get_frame(src->rx_handle);
+    if (frame) {
+      break;
+    }
+  }
+
+  if (!frame) {
+    GST_INFO("Failed to get frame EOS");
+    GST_OBJECT_UNLOCK(src);
+    return GST_FLOW_EOS;
+  }
+
+  gst_buffer_map(buf, &dest_info, GST_MAP_WRITE);
+  fill_size = gst_buffer_fill(buf, 0, frame->addr[0], src->frame_size);
+  gst_buffer_unmap(buf, &dest_info);
+
+  if (fill_size != src->frame_size) {
+    GST_ERROR("Failed to fill buffer");
+    ret = GST_FLOW_ERROR;
+  } else {
+    ret = GST_FLOW_OK;
+  }
+
+  st20p_rx_put_frame(src->rx_handle, frame);
```

Comment:
It will send buffer even if it fails to fill it.. Are we ok with it?   ( I'm not sure if we can return the buffer without sending it)